### PR TITLE
Reachability slicer: split cfg walks into two phases

### PR DESCRIPTION
**Pull request by smowton**
The first phase walks outwards towards __CPROVER__start, visiting all callers but recording
callsites seen en route, then the second phase walks into callees not seen in the first phase,
restricting the search to known reachable functions. This fixes the previous algorithm, which
attempted the same logic but could visit too few functions when a particular functions was
encountered as a callee before later being seen as a caller -- the latter results in more code
being visited because the calling context is unknown, and so must be done first.

This is an alternative to #2748, see comments there for more detail.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 4d1cc88151ad36880929eb3627ab92d71efdd8f8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88799021

File: src/goto-instrument/reachability_slicer.cpp [hunk 1]
```diff
@@ -138,51 +240,81 @@ void reachability_slicert::fixedpoint_from_assertions(
...
 
+/// Perform forwards depth-first search of the control-flow graph of the
+/// goto program, starting from the nodes corresponding to the criterion and
+/// the instructions that might be executed concurrently. Set reaches_assertion
+/// to true for every instruction visited.
+/// \param is_threaded Instructions that might be executed concurrently
+/// \param criterion the criterion we are trying to hit
+void reachability_slicert::fixedpoint_from_assertions(
+  const is_threadedt &is_threaded,
+  slicing_criteriont &criterion)
```
Comment (peterschrammel) [inline src/goto-instrument/reachability_slicer.cpp]: what are the side effects on this?

File: regression/cbmc/reachability-slice-interproc2/test.desc [hunk 2]
```diff
@@ -0,0 +1,8 @@
+CORE
+main.c
+--reachability-slice --show-goto-functions
+^EXIT=0$
+^SIGNAL=0$
+main\(\)
+--
+^warning: ignoring
```
Comment (peterschrammel) [inline regression/cbmc/reachability-slice-interproc2/test.desc:8]: Can you check that `b` is not in the slice?

Comment (smowton) [reply regression/cbmc/reachability-slice-interproc2/test.desc:8]: No, it is in the slice, since it calls c.

Comment (peterschrammel) [reply regression/cbmc/reachability-slice-interproc2/test.desc:8]: Ah, I see, the forward walk is not used in the way I assumed.

Comment (peterschrammel) [reply regression/cbmc/reachability-slice-interproc2/test.desc:8]: Is there a test that actually slices away something?

File: src/goto-instrument/reachability_slicer.cpp (hunk 1 above)
Comment (smowton) [reply src/goto-instrument/reachability_slicer.cpp]: None. Pushed an extra commit to constify these, at the risk of @tautschnig's ire re: scope creep ;)

File: regression/cbmc/reachability-slice-interproc2/test.desc (hunk 2 above)
Comment (smowton) [reply regression/cbmc/reachability-slice-interproc2/test.desc:8]: The original cbmc/reachability-slice has two tests looking for slicing. These new ones were particularly looking for over-slicing since that's the bug @polgreen / @tautschnig encountered.

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 0d8e3c1407ab42202effb387373e06eb4b6a04eb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88837602

**Comment by tautschnig** [comment]
As a side remark before getting to the key parts: "Constify slicing_criteriont::operator()" would make for a nice separate PR - it could likely be merged more or less immediately, but also it may need to be reverted in future in case someone comes up with a criterion that evolves as you query it (maybe that's weird, but that was my thinking when I originally wrote this code and I thus left it non-const).

File: src/goto-instrument/reachability_slicer.cpp (hunk 1 above)
Comment (tautschnig) [reply src/goto-instrument/reachability_slicer.cpp]: There you go :-)

**Review by tautschnig — APPROVED** [review]
I think this is ok, but I can hardly believe we'd be the first to solve this problem. @kroening mentioned BEBOP (I think that would be Fig. 3 of https://drona.csa.iisc.ac.in/~deepakd/tpa-2016/bebop.pdf), and there might be others. If at all possible, we should be re-using algorithms and terminology and refer to those for documentation.

**Comment by smowton** [comment]
I find it quite hard to parse those academic code listings with highly abbreviated and subscript-littered function and variable names, but I don't *think* this particularly relates to their reachability algorithm, because the algorithm described in Figure 3 starts at the head of `main` and works forwards, meaning it should never deal with a return edge whose callee is unknown. Therefore it won't face our problem, that a given program point can be encountered while walking outwards (off the top of the notional stack) and while walking inwards (with a stack).

**Comment by smowton** [comment]
I'll have a search for something comparable though, and let you know what I find. Merging these meanwhile, then will follow up with a doc PR if I find anything.
